### PR TITLE
Use xxHash to implement StableHasher

### PR DESCRIPTION
**Pull request by Aaron1011**

### CODE REVIEW

File: compiler/rustc_data_structures/src/stable_hasher.rs [hunk 1]
```diff
@@ -32,7 +32,7 @@ pub trait StableHasherResult: Sized {
 impl StableHasher {
     #[inline]
     pub fn new() -> Self {
-        StableHasher { state: SipHasher128::new_with_keys(0, 0) }
+        StableHasher { state: Hash128::with_seed(0) }
```
Comment (nagisa) [inline compiler/rustc_data_structures/src/stable_hasher.rs:35]: Passing thought, should we perhaps consider using some other constant than 0 for the initial seed? Some hashers are known to produce pretty poor results when the initial seed is 0, which might not apply for these specific hash algorithms, but specifying a non-0 value would likely obviate concerns like these regardless.

**Comment by bstrie** [comment]
The last perf run didn't look too promising, is this still worth pursuing? What initial problem is this seeking to solve?

File: compiler/rustc_data_structures/src/stable_hasher.rs (hunk 1 above)
Comment (thomcc) [reply compiler/rustc_data_structures/src/stable_hasher.rs:35]: This should use `State::default()`, or a lot of redundant work will be done.

**Review by thomcc — COMMENTED** [review]
IIRC the twox crate isn't a particularly high performance implementation of the xxh algorithms. I don't know if it's just based on an old version of xxh or what, but there are a lot of things it does that will lead to much worse performance than the C/reference implementation. A couple examples are:

- Using heap allocation here https://github.com/shepmaster/twox-hash/blob/main/src/xxh3.rs#L855 ([compare to ref/C impl](https://github.com/Cyan4973/xxHash/blob/be534ed8c9fa677f18f230735c114373b1ed7d34/xxhash.h#L987)), means means that the loads performed in the SIMD accel versions need to be unaligned, which has *wildly* varying latencies on SIMD. (Also, I suspect that the optimizer can't see that it's not actually resizing the buffer a bunch)
- I highly suspect the way the `With` enum internally will lead to the same thing, and branching at each iteration. (I think the body of these loops are too bug for LLVM's loop unswitching to kick in)

There are probably others, this was just a glance I did when I found the crate to be about 3x slower than the C code. I ended up using it anyway since it meant I could move that code from C++ to rust, but I was pretty disappointed. (Probably, I should have filed issues, but at the time I was very busy)

Another thing is that for small keys, XXH's api assumes you aren't using the streaming API. but the way the `core::hash` API is set up, the optimizations it does like that are much more difficult. I don't know if that's relevant here, though.

**Comment by Mark-Simulacrum** [comment]
Going to go ahead and close this, since it seems unlikely to get directly picked back up again.

**Comment by thomcc** [comment]
Is there any information on the set of properties needed/desired for StableHasher? Is it just "collisions rare in practice"? (xxHash collisions can be forced, for example)

**Comment by bjorn3** [comment]
Collisions must be highly unlikely even in adversarial cases. A hash collision for `StableHasher` will almost always result in a miscompilation, unlike for `Hash` where it will just reduce performance.

**Comment by thomcc** [comment]
Okay, that's mostly what I thought. Algorithms like xxHash cannot be used then.

**Pull request closed without merging**